### ISSUE TITLE
bloc_test does not show what is exactly different when a test fails. flutter test does

### ISSUE THREAD

**Issue by ride4sun** [felangel/bloc#1313]
**Is your feature request related to a problem? Please describe.**
I have a test that failed which has a lot of output and only a small section is different compared to the expected output.  I think flutter test shows exactly what is different between expected and actual. bloc_test lets me compare the results by staring at it. 

**Describe the solution you'd like**
A simple diff would be great. Maybe show the diff only when the difference is not bigger than a threshold.

**Describe alternatives you've considered**
Observe by looking at it which is time-consuming

**Additional context**
This is one of my failing test which would be greatly improved if I would see a diff:

```
Expected: [
            SendTest<dynamic>:SendTest(request:Commands: [atd, atkw0, atsp0, ate0, ati, ath1, ats0, atl0, atat1, 0100, 0902, atdp] continuous: false id: setup, updateRate: 0 }),
            Analyzed<dynamic>:Analyzed(request:Commands: [atd, atsp0, atkw0, ate0, ath1, atal, ati, ats0, atl0, atdp, 0100, 0900, 0902, atdp] continuous: false id: setup, updateRate: 0 }),
            Finished<CommunicationInitResult>:Finished(results:DateTime:  2020-06-16 05:02,   
           Protocol: ISO9141-2,   
           Selected ECU: 10,   
          Valid ECUs: [10, 1A],   
          PID Support 100: [[10, BE3EB811], [1A, 98188001]],   
          Header Filtering: true,   
          Vin: null,   
          first request/response: [atd, atkw0, atsp0, ate0, ati, ath1, ats0, atl0, atat1, 0100, 0902, atdp],   
          modified request: [atd, atsp0, atkw0, ate0, ath1, atal, ati, ats0, atl0, atdp, 0100, 0900, 0902, atdp],  
           final response:  [[atd, atdOK], [atsp0, atsp0OK], [atkw0, atkw0OK], [ate0, OK], [ath1, OK], [atal, OK], [ati, ELM327v15], [ats0, OK], [atl0, OK], [atdp, AUTO,ISO9141-2], [0100, 486B104100BE3EB811C9486B1A4100981880013F], [0900, NODATA], [0902, NODATA], [atdp, AUTO,ISO9141-2]],   
           error Description:  no error,   
           )
          ]
  Actual: [
            SendTest<dynamic>:SendTest(request:Commands: [atd, atkw0, atsp0, ate0, ati, ath1, ats0, atl0, atat1, 0100, 0902, atdp] continuous: false id: setup, updateRate: 0 }),
            Analyzed<dynamic>:Analyzed(request:Commands: [atkw0, ate0, ath0, atal, ati, ats0, atl0, atsh6C10F1, 0100, 0900, 0902, atdp] continuous: false id: setup, updateRate: 0 }),
            Finished<CommunicationInitResult>:Finished(results:DateTime: 2020-06-16 05:50,   
           Protocol: ISO9141-2,   
           Selected ECU: 10,   
          Valid ECUs: [10, 1A],   
          PID Support 100:   
          10 : BE3EB811,   
          1A : 98188001,   
            
          ,   
          Header Filtering: true,   
          Vin: null,   
          first request/response:   
          atd : atdOK,   
          atkw0 : atkw0OK,   
          atsp0 : atsp0OK,   
          ate0 : ate0OK,   
          ati : ELM327v15,   
          ath1 : OK,   
          ats0 : OK,   
          atl0 : OK,   
          atat1 : OK,   
          0100 : 486B104100BE3EB811C9486B1A4100981880013F,   
          0902 : NODATA,   
          atdp : AUTO,ISO9141-2,   
            
          ,   
          modified request: [atkw0, ate0, ath0, atal, ati, ats0, atl0, atsh6C10F1, 0100, 0900, 0902, atdp],  
           final response:    
          atd : atdOK,   
          atsp0 : atsp0OK,   
          atkw0 : atkw0OK,   
          ate0 : OK,   
          ath1 : OK,   
          atal : OK,   
          ati : ELM327v15,   
          ats0 : OK,   
          atl0 : OK,   
          atdp : AUTO,ISO9141-2,   
          0100 : 486B104100BE3EB811C9486B1A4100981880013F,   
          0900 : NODATA,   
          0902 : NODATA,   
          atdp : AUTO,ISO9141-2,   
            
          ,   
           error Description:  ,   
           )
          ]
   Which: was Analyzed<dynamic>:<Analyzed(request:Commands: [atkw0, ate0, ath0, atal, ati, ats0, atl0, atsh6C10F1, 0100, 0900, 0902, atdp] continuous: false id: setup, updateRate: 0 })> instead of Analyzed<dynamic>:<Analyzed(request:Commands: [atd, atsp0, atkw0, ate0, ath1, atal, ati, ats0, atl0, atdp, 0100, 0900, 0902, atdp] continuous: false id: setup, updateRate: 0 })> at location [1]

package:test_api                             expect
package:bloc_test/src/bloc_test.dart 143:29  blocTest.<fn>.<fn>
===== asynchronous gap ===========================
dart:async                                   _asyncThenWrapperHelper
package:bloc_test/src/bloc_test.dart         blocTest.<fn>.<fn>
dart:async                                   runZoned
package:bloc_test/src/bloc_test.dart 135:11  blocTest.<fn>

```

**Comment by felangel** [comment]
Hi @ride4sun 👋 
Thanks for opening an issue!

Under the hood `blocTest` uses `test` from `package:test` and calls `expect` on the states that were emitted vs the expected states.

https://github.com/felangel/bloc/blob/aef1104e33009439cbd0a419afb000f1d7573635/packages/bloc_test/lib/src/bloc_test.dart#L143

I will look into ways to customize the diffing but I believe the behavior is consistent with what it would be if you had just written a regular flutter or dart test.

You can check this by writing a regular test:

```dart
test('state diffing', () {
  final a = [...]; // list of states
  final b = [...]; // slightly different list of state
  expect(a, b); // should fail and you should see the same diff as with blocTest
});
```

Is there an example of the improved diffing that you can share? Thanks! 🙏

**Comment by felangel** [comment]
@ride4sun thoughts on my previous comment?

**Comment by ride4sun** [comment]
Well, I am looking for a better indication of what is different between the Actual and the Expected. In my example above there is no indication besides that they are different. That makes it hard to inspect the outcome fast. It would be nice if there would be something like that:

```Actual was  => ...modified request: [atkw0, ate0, ath0, atal, ati, ats0, atl0, atsh6C10F1, ...```
```Expected is => ...modified request: [atkw0, ate0, ath1, atal, ati, ats0, atl0, atsh6C10F2, ... ```
                                                                                      ^                                               ^

It does not have to be a complete site by site diff output. Just the beginning of the difference would be enough to move on quickly. 

Does that make sense.

**Comment by ride4sun** [comment]
Actualy I found an direct example: 
Here is the test:
```
  test(
      'Test VinDecoder Simulator 9541 default setup '
      '[0902, 486B104902010000003140486B1049020247314A4315486B1049020335343434E2486B104902045237323502486B1049020532333637E5] ',
      () {
    model.VinDecoder decoder = model.VinDecoder();
    decoder.decode(
        '486B104902010000003140486B1049020247314A4315486B1049020335343434E2486B104902045237323502486B1049020532333637E5',
        (_) {},
        UnitSystem.Metrics);
    expect(decoder.isValid, equals(true));
//    expect(decoder.manufacturer, equals('Toyota'));

    expect(decoder.vin, equals('1D4GP00R55B123456'));
  });
```

Here is the output diff

```
Expected: '1D4GP00R55B123456'
  Actual: '1G1JC5444R7252367'
   Which: is different.
          Expected: 1D4GP00R55B ...
            Actual: 1G1JC5444R7 ...
                     ^
           Differ at offset 1
```

**Comment by felangel** [comment]
@ride4sun thanks will take a closer look at this soon 👍

**Comment by AlexKolykhalov** [comment]
Hi, @felangel, in my case i am just trying to get output with differences when i change code from https://bloclibrary.dev/#/testing
``` 
blocTest(
    'emits [1] when CounterEvent.increment is added',
    build: () => counterBloc,
    act: (bloc) => bloc.add(CounterEvent.increment),
    expect: [1],
);

```
to

```
blocTest(
    'emits [1] when CounterEvent.increment is added',
    build: () => counterBloc,
    act: (bloc) => bloc.add(CounterEvent.increment),
    expect: [2],
);

```
but i get TimeoutException after 0:00:30.000000: Test timed out after 30 seconds

**Comment by felangel** [comment]
@ride4sun closing this because `blocTest` behaves the same as a flutter test -- the difference here is you are individually expecting individual properties are equal to an expected value whereas in blocTest you are providing entire state objects. I believe this is solved by #1275 because we will provide a warning to the user to verify that the states are comparable and if you want additional information (and are using Equatable) you can set `stringify` to `true` to get a complete debug print of the state 👍 

Hope that helps and feel free to comment with additional info/thoughts if you feel the issue is not resolved, thanks!

**Comment by yvonmanzi** [comment]
@felangel thanks for the explanation. I was running into a similar issue because my state class was not implementing Equatable. Thus there was no way to actually compare the states.
